File: multiple languages/1026. Maximum Difference Between Node and Ancestor/main.ts
/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     val: number
 *     left: TreeNode | null
 *     right: TreeNode | null
 *     constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.left = (left===undefined ? null : left)
 *         this.right = (right===undefined ? null : right)
 *     }
 * }
 */

 const maxAncestorDiff = (root: TreeNode | null): number => {

    const dfs = (node:TreeNode,min_node_value:number, max_node_value:number) : number => {

        if (!node) {
            return Math.abs(min_node_value-max_node_value);
        }
        min_node_value = Math.min(min_node_value, node.val)
        max_node_value = Math.max(max_node_value, node.val)

        return Math.max(dfs(node.left,min_node_value,max_node_value),
        dfs(node.right,min_node_value,max_node_value))

    }
    return dfs(root,root.val,root.val);
};

File: multiple languages/1026. Maximum Difference Between Node and Ancestor/main.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func maxAncestorDiff(root *TreeNode) int {
	return dfs(root, root.Val, root.Val)
}

func dfs(node *TreeNode, min_node_value int, max_node_value int) int {

	if node == nil {
		return int(math.Abs(float64(min_node_value - max_node_value)))
	}

	min_node_value = min(min_node_value, node.Val)
	max_node_value = max(max_node_value, node.Val)

	return max(dfs(node.Left, min_node_value, max_node_value),
		dfs(node.Right, min_node_value, max_node_value))

}

func max(i, j int) int {
	if i < j {
		return j
	}
	return i
}

func min(i, j int) int {
	if i > j {
		return j
	}
	return i
}
